### DISCUSSION TITLE
Compare 2 pdf´s and show difference with 2 colors

### DISCUSSION THREAD

**Discussion by Michipapa** [ImageMagick/ImageMagick#2233] [Help]
Hi,

I found this wonderful tool to compare 2 technical drawings (pdf). It works like a charme.
One question: The result pdf shows all diffs in red, but I want:

Things who are in pdf2 and not in pdf1 in red 
and
Things who are in pdf1 and not in pdf2 in blue.

Is that possible ?

regards Michael

**Comment by fmw42** [comment]
What was your exact command line?  Can you post your two input images? What is your IM version and platform/OS?

**Reply by Michipapa** [reply]
magick compare -density 150 1.pdf 2.pdf -fuzz 10%  delta-1_2.pdf

[delta-1_2.pdf](https://github.com/ImageMagick/ImageMagick/files/4873901/delta-1_2.pdf)
[2.pdf](https://github.com/ImageMagick/ImageMagick/files/4873902/2.pdf)
[1.pdf](https://github.com/ImageMagick/ImageMagick/files/4873903/1.pdf)

**Reply by Michipapa** [reply]
PS: 7.0.10 (downloaded today) and Windows 10 64 bit

regards Michael

**Reply by fmw42** [reply]
Here are two ways. I was not sure what you wanted for the result.

Unix Syntax:

```
magick -density 150 1.pdf 2.pdf \
\( -clone 0 -fill white -colorize 80%  \) \
\( -clone 0,1 -compose minus -composite -fuzz 20% -fill red -opaque white -transparent black \) \
\( -clone 1,0 -compose minus -composite -fuzz 20% -fill blue -opaque white -transparent black \) \
-delete 0,1 \
-background white -compose over -flatten \
12a.pdf
```

[12a.pdf](https://github.com/ImageMagick/ImageMagick/files/4874253/12a.pdf)




```
magick -density 150 1.pdf 2.pdf \
\( -clone 0,1 -compose minus -composite -fuzz 20% -fill red -opaque white -transparent black \) \
\( -clone 1,0 -compose minus -composite -fuzz 20% -fill blue -opaque white -transparent black \) \
-delete 0,1 \
-background white -compose over -flatten \
12b.pdf
```

[12b.pdf](https://github.com/ImageMagick/ImageMagick/files/4874255/12b.pdf)


Windows Syntax:

```
magick -density 150 1.pdf 2.pdf ^
( -clone 0 -fill white -colorize 80% ) ^
( -clone 0,1 -compose minus -composite -fuzz 20% -fill red -opaque white -transparent black ) ^
( -clone 1,0 -compose minus -composite -fuzz 20% -fill blue -opaque white -transparent black ) ^
-delete 0,1 ^
-background white -compose over -flatten ^
12a.pdf
```

```
magick -density 150 1.pdf 2.pdf ^
( -clone 0,1 -compose minus -composite -fuzz 20% -fill red -opaque white -transparent black ) ^
( -clone 1,0 -compose minus -composite -fuzz 20% -fill blue -opaque white -transparent black ) ^
-delete 0,1 ^
-background white -compose over -flatten ^
12b.pdf
```

**Reply by Michipapa** [reply]
Hi,

thx for your code.

It works with my simple example, but not with the real complex drawings (no blue part). Unfortunately I can´t upload them cause of security issues.

regards Michael

**Comment by snibgo** [comment]
A minor point about fmw42's Windows commands: if you use them in a BAT script, you need to double the percent % signs, eg "20%%".

**Comment by Michipapa** [comment]
Hi snibgo,

thx for your comment. I change this in my batch but no difference in the result :-(

regards Michael

**Comment by snibgo** [comment]
> It works with my simple example, but not with the real complex drawings (no blue part). 

In what way does it not work?

You say you want ...
> Things who are in pdf2 and not in pdf1 in red

... but how is that defined? IM can easily find what pixels are different between two images. fmw42's commands assume you have dark "things" on a white background. But that assumption may not be correct.

**Comment by Michipapa** [comment]
Hi snibgo,

look at my examples at the top of the thread.

we have technical drawings that are updated in the course of development. Each drawing differs a bit from its previous version. Every version gets a so-called "index". The drawings are very complex so that the viewer cannot immediately see what is new and what is gone for each index.
That's why he needs a tool that e.g. marked in color

gray=the same in both versions ("index")
red = new
blue = no longer exists

IMHO, the tool must compare pdf1 with pdf2 and mark all differences in red, then pdf2 with pdf1 and mark it in blue. and finaly put it together. 

regards Michael

regards Michael

**Comment by snibgo** [comment]
>IMHO, the tool must compare pdf1 with pdf2 and mark all differences in red, then pdf2 with pdf1 and mark it in blue. and finaly put it together.

When we compare a pixel in one image with the corresponding pixel in another image, then either the pixels are the same or they are different.

The order of the comparison makes no difference.

**Reply by Michipapa** [reply]
if the background is the reference color (=white), I can tell:

Pixel at position 10,10 was in pdf1 = background and is no longer in pfd2 ==> display it red in diff

Pixel at position 20.20 was in pdf1 <> background and is now in pdf2 ==> display it blue in diff

Or am I wrong?

**Comment by snibgo** [comment]
That is what fmw42's commands do. But you say they don't work for your complex drawings. What goes wrong?

**Comment by fmw42** [comment]
Can you make an example or cut a small pieces from your PDF files that show this issue? Are you PDF files transparent background? Are they CMYK or RGB?

**Comment by Michipapa** [comment]
@snibgo :  There is no blue part in diff page
@fmw42 : Yes, I prepair a sample tomorrow.

regards michael

**Reply by fmw42** [reply]
What was your exact command? Were you trying to do it in mogrify or convert?

**Comment by snibgo** [comment]
Perhaps you want the blue rectangle and arrow and triangle to be in the result? Well, fmw42's command removes both inputs from the output, so the output contains only white (no change) or red (only in pdf1) or blue (only in pdf2). That's what you said you want. But if you also want the blue shapes in the output then change ...
```
-delete 0,1
```
... to ...
```
-delete 0
```

**Comment by Michipapa** [comment]
Ok, here are some parts of a real technical drawings.

[21_part.pdf](https://github.com/ImageMagick/ImageMagick/files/4877481/21_part.pdf)
[22_part.pdf](https://github.com/ImageMagick/ImageMagick/files/4877482/22_part.pdf)
[g-diff.pdf](https://github.com/ImageMagick/ImageMagick/files/4877483/g-diff.pdf)
[m-diff.pdf](https://github.com/ImageMagick/ImageMagick/files/4877484/m-diff.pdf)

21_part.pdf (lower index of the drawing)
22_part.pdf (higher index of the drawing)

g_diff.pdf (result from GraphicsMagics with the command gm compare %1 %2 -density 300  -highlight-color red -type grayscale -file %3  )
As you can see, it shows new elements in 22 in darkred and old elements from 21 in lightred. Near perfect, if I could replace lightred with blue to better recognisation. And unfortunately it ignors the -density command so large pdfs have a low resolution.

m-diff.pdf is the result from ImageMagics compare -density 150 %1 %2 -fuzz 20%%   -highlight-color red %3 

regards Michael

**Comment by snibgo** [comment]
> Michipapa: if the background is the reference color (=white), ...

That is not true. Your files 21_part.pdf and 22_part.pdf do not have a white background. They have a transparent background. We can flatten each against white.
```
magick ^
  -background White -density 150 ^
  ( 21_part.pdf -layers Flatten ) ^
  ( 22_part.pdf -layers Flatten ) ^
  ( -clone 0,1 -compose minus -composite -fuzz 20%% -fill red -opaque white -transparent black +write d1.png ) ^
  ( -clone 1,0 -compose minus -composite -fuzz 20%% -fill blue -opaque white -transparent black +write d2.png ) ^
  -delete 0 ^
  -background white -compose over -flatten ^
12b.png
```
I have saved intermediate results d1.png and d2.png. From these, we can clearly see that some elements have moved slightly between the images, so they show as both additions and deletions.

"-density X" should be before reading the PDF inputs, not after.

**Reply by Michipapa** [reply]
Hi, snibo.

You are my hero :-)

Wow,  works really good with the original files. And many thx for the hint with -density, thats was the cause ...

Last (blasphemic) question,

because it takes some minutes and consume a lot of memory:

Would these also possible with "GraphicMagics" ? There goes it faster and need not so much RAM.

The commands similar, but not identical.

regards Michael

**Reply by liddell-d** [reply]
I have a related question.

Here is a slightly modified version of the code shown above:

```
convert \
  -background white -density 150 \
  '(' a.pdf -flatten ')' \
  '(' b.pdf -flatten ')' \
  '(' -clone 0,1 -compose minus -composite -fuzz 20% -fill red  -opaque white -transparent black +write diff_01.pdf ')' \
  '(' -clone 1,0 -compose minus -composite -fuzz 20% -fill blue -opaque white -transparent black +write diff_10.pdf ')' \
  -delete 0 \
  -background white -compose over -flatten \
diff.pdf
```

Here are the input files:

[a.pdf](https://github.com/ImageMagick/ImageMagick/files/15384600/a.pdf)
[b.pdf](https://github.com/ImageMagick/ImageMagick/files/15384601/b.pdf)

Here is the output:

[diff.pdf](https://github.com/ImageMagick/ImageMagick/files/15384612/diff.pdf)

Both the output and the intermediate diff_*.pdf files created by the script are entirely black. I don't know why. Can anyone suggest a fix to the code that will show the difference in the page numbers?

**Reply by snibgo** [reply]
Your inputs a.pdf and b.pdf are encoded as CMYK. This changes the meaning of colour names like "red2", "blue" and "white". I suggest you tell IM to rasterize the PDF input as sRGB:
```
magick \
  -background white -density 150 \
  '(' -colorspace sRGB a.pdf -flatten ')' \
  '(' -colorspace sRGB b.pdf -flatten ')' \
  '(' -clone 0,1 -compose minus -composite -fuzz 20% -fill red  -opaque white -transparent black +write diff_01.pdf ')' \
  '(' -clone 1,0 -compose minus -composite -fuzz 20% -fill blue -opaque white -transparent black +write diff_10.pdf ')' \
  -delete 0 \
  -background white -compose over -flatten \
  -write diff.png \
diff.pdf
```

**Comment by snibgo** [comment]
Sorry, I don't know GraphicsMagick.
